substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java

```diff
@@ -125,75 +134,92 @@ public boolean openFile(String outputFile) {
-        Target_java_nio_DirectByteBuffer bb = new Target_java_nio_DirectByteBuffer(JfrBufferAccess.getDataStart(buffer).rawValue(), capacity);
-        FileChannel fc = file.getChannel();
-        try {
-            fc.write(SubstrateUtil.cast(bb, ByteBuffer.class));
-            return file.getFilePointer() > notificationThreshold;
-        } catch (IOException e) {
-            Logger.log(JFR_SYSTEM, ERROR, "Error while writing file " + filename + ": " + e.getMessage());
+        assert (JfrBufferAccess.isAcquired(buffer) || VMOperation.isInProgressAtSafepoint());
+        UnsignedWord unflushedSize = JfrBufferAccess.getUnflushedSize(buffer);
+        if (unflushedSize.equal(0)) {
             return false;
         }
+
+        boolean success = getFileSupport().write(fd, buffer.getTop(), unflushedSize);
+        if (!success) {
+            return false;
+        }
+        JfrBufferAccess.increaseTop(buffer, unflushedSize);
+        return getFileSupport().position(fd).greaterThan(WordFactory.signed(notificationThreshold));
     }
 
     /**
      * We are writing all the in-memory data to the file. However, even though we are at a
      * safepoint, further JFR events can still be triggered by the current thread at any time. This
-     * includes allocation and GC events. Therefore, it is necessary that our whole JFR
-     * infrastructure is epoch-based. So, we can uninterruptibly switch to a new epoch before we
-     * start writing out the data of the old epoch.
+     * includes allocation and GC events. Therefore, it is necessary that we switch
+     * to a new epoch in an uninterruptible safepoint
      */
     // TODO: add more logic to all JfrRepositories so that it is possible to switch the epoch. The
     // global JFR memory must also support different epochs.
     public void closeFile(byte[] metadataDescriptor, JfrRepository[] repositories) {
         assert lock.isHeldByCurrentThread();
-        JfrCloseFileOperation op = new JfrCloseFileOperation(metadataDescriptor, repositories);
+        JfrCloseFileOperation op = new JfrCloseFileOperation();
         op.enqueue();
+
+        // JfrCloseFileOperation will switch to a new epoch so data for the old epoch will not
+        // be modified by other threads and can be written without a safepoint
+        try {
+            SignedWord constantPoolPosition = writeCheckpointEvent(repositories);
+            SignedWord metadataPosition = writeMetadataEvent(metadataDescriptor);
+            patchFileHeader(constantPoolPosition, metadataPosition);
+            getFileSupport().close(fd);
+        } catch (IOException e) {
+            Logger.log(JFR_SYSTEM, ERROR, "Error while writing file " + filename + ": " + e.getMessage());
+        }
+
+        filename = null;
+        if (shouldNotify) {
+            shouldNotify = false;
+            //Checkstyle: stop
+            synchronized (Target_jdk_jfr_internal_JVM.FILE_DELTA_CHANGE) {
```

Comment:
I am wondering if there are still deadlocks possible, i.e., is there any (JDK) code that could acquire `JfrChunkWriter.lock` while holding `Target_jdk_jfr_internal_JVM.FILE_DELTA_CHANGE`?

Actually, it might be possible to completely remove this code. We already closed the file, so there shouldn't be any need to do the notify. At least I don't think that HotSpot does notify in this particular situation.

Reply:
I reviewed the notify mechanism; I think Victor is looking at revising it to be similar to HotSpot and you're right, the notify shouldn't be here, so it's removed now.